walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,117 @@
+      TimeRange previous = takenTimes.get(0);
+
+      for (TimeRange time : takenTimes) {
+        if (time.overlaps(previous)) {
+          int start = previous.start();
+          int end = Math.max(time.end(), previous.end());
+          previous = TimeRange.fromStartEnd(start, end, false);
+        } else {
+          takenTimesMerged.add(previous);
+          previous = time;
+        }
+      }
+      takenTimesMerged.add(previous);
+    }
+    return takenTimesMerged;
+  }
+
+  private ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged) {
+    int END_OF_DAY = 1440;
+    int START_OF_DAY = 0;
+    TimeRange emptyAtStart = TimeRange.fromStartDuration(START_OF_DAY,0);
+    TimeRange emptyAtEnd = TimeRange.fromStartDuration(END_OF_DAY, 0);
+    takenTimesMerged.add(0, emptyAtStart);
+    takenTimesMerged.add(emptyAtEnd);
+    
+    ArrayList<TimeRange> validTimes = new ArrayList();
+    for (int i = 0; i < takenTimesMerged.size() - 1; i++) {
+      TimeRange currentRange = takenTimesMerged.get(i);
+      TimeRange adjRange = takenTimesMerged.get(i + 1);
+      int newStart = currentRange.end();
+      int newEnd = adjRange.start();
+      validTimes.add(TimeRange.fromStartEnd(newStart, newEnd, false));
+    }
+
+    System.out.println(validTimes);
+    return validTimes;
+  }
+
+  private ArrayList<TimeRange> filterTimes(ArrayList<TimeRange> validTimes, MeetingRequest request) {
+    ArrayList<TimeRange> finalTimes = new ArrayList();
+    for (TimeRange time : validTimes) {
+      if (time.duration() >= request.getDuration()) {
+          finalTimes.add(time);
+      }
+    }
+    return finalTimes;
   }
+
+
+  /**
+   * Parameters: a collection of events occuring that day, and a meeting request containings
+   * a time for the meeting, duration, and attendence list of manadatory and optional attenedees
+   * a time for the meeting,
+   * 
+   * Response: Based on the events happeing on the day, query will find and return time ranges
+   * in which the meeting can be hosted because no mandatory members will have a conflict.
+   * It will try to factor in optional attendees so long as they do not cause the mandatory
+   * members to not attend. 
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+    if (events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    ArrayList<TimeRange> takenTimes = new ArrayList();
+    ArrayList<TimeRange> takenTimesOptional = new ArrayList();
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+          takenTimes.add(event.getWhen());
+        }
+      if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+          takenTimesOptional.add(event.getWhen());
+        }
+    }
+
+    if (!takenTimes.isEmpty()) {
```

Comment:
What you have here is fine but it looks like you could simplify this a bit if you want since the first 3 lines of each branch are duplicated